### PR TITLE
[NUI] Introduce NUI Palette APIs

### PR DESCRIPTION
**Pull request by lwc0917**
### Description of Change ###
Introduce NUI Palette APIs

All the detail about it
https://code.sec.samsung.net/confluence/display/ENUIFWC/94.+NUI+Palette

### API Changes ###
No public APIs yet

### CODE REVIEW

**Comment by JoogabYun** [comment]
Please use Pascal Case for a static public/protected field name and a API name.
do not append prefix 'm' like 'mColors'.

Please refer below site.
https://code.sec.samsung.net/confluence/pages/viewpage.action?pageId=191373542

**Comment by lwc0917** [comment]
@JoogabYun I've updated commit to following your guide. Thanks

File: src/Tizen.NUI/src/internal/Utility/ColorCutQuantizer.cs [hunk 1]
```diff
@@ -0,0 +1,633 @@
...
+using System.Drawing;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Tizen.NUI
+{
+    internal sealed class ColorCutQuantizer
+    {
+        private const float BlackMaxLightness = 0.05f;
```
Comment (JoogabYun) [inline src/Tizen.NUI/src/internal/Utility/ColorCutQuantizer.cs:34]: private variables have to start with a lowercase letter.
'BlackMaxLightness' -> 'blackMaxLightness'

please check other private variables as well

File: src/Tizen.NUI/src/internal/Utility/ColorCutQuantizer.cs [hunk 2]
```diff
@@ -0,0 +1,633 @@
...
+            {
+                throw new ArgumentNullException(nameof(colorHistogram), "colorHistogram should not be null.");
+            }
+
+            if (maxColors < 1)
+            {
+                throw new ArgumentNullException(nameof(maxColors), "maxColors should not be null.");
+            }
+
+            int RawColorCount = colorHistogram.GetNumberOfColors();
```
Comment (JoogabYun) [inline src/Tizen.NUI/src/internal/Utility/ColorCutQuantizer.cs:57]: please start with a lowercase latter.
'RawColorCount' -> 'rawColorCount'

please check other variables as well

File: src/Tizen.NUI/src/internal/Utility/ColorCutQuantizer.cs [hunk 3]
```diff
@@ -0,0 +1,633 @@
...
+            int[] RawColors = colorHistogram.GetColors();
+            int[] RawColorCounts = colorHistogram.GetColorCounts();
+
+            // First, lets pack the populations into a SparseIntArray so that they can be easily
+            // retrieved without knowing a color's index
+            ColorPopulations = new Dictionary<int, int>();
+
+            for (int i = 0; i < RawColors.Length; i++)
+            {
+                System.Drawing.Color RgbColor = System.Drawing.Color.FromArgb(RawColors[i]);
```
Comment (JoogabYun) [inline src/Tizen.NUI/src/internal/Utility/ColorCutQuantizer.cs:67]: 'RgbColor' -> 'rgbColor'

File: src/Tizen.NUI/src/internal/Utility/ColorCutQuantizer.cs [hunk 4]
```diff
@@ -0,0 +1,633 @@
...
+                QuantizedColors = QuantizePixels(ValidColorCount - 1, maxColors);
+            }
+        }
+
+        /// <summary>
+        /// Factory-method to generate a ColorCutQuantizer from a  Bitmap object.
+        /// </summary>
+        public static ColorCutQuantizer FromBitmap(Bitmap bitmap, Tizen.NUI.Rectangle region, int maxColors)
+        {
+            int Width;
```
Comment (JoogabYun) [inline src/Tizen.NUI/src/internal/Utility/ColorCutQuantizer.cs:107]: 'Width' -> 'width'

File: src/Tizen.NUI/src/internal/Utility/ColorCutQuantizer.cs [hunk 5]
```diff
@@ -0,0 +1,633 @@
...
+                    Colors.Add(Color);
+                }
+            }
+            Tizen.Log.Info("Palette", "Final generated color count = " + Colors.Count + "\n");
+            return Colors;
+        }
+
+        private Boolean ShouldIgnoreColor(int color)
+        {
+            System.Drawing.Color C = System.Drawing.Color.FromArgb(color);
```
Comment (JoogabYun) [inline src/Tizen.NUI/src/internal/Utility/ColorCutQuantizer.cs:206]: Please use meaningful naming.
'C' -> 'drawingColor' etc...

File: src/Tizen.NUI/src/internal/Utility/ColorCutQuantizer.cs [hunk 6]
```diff
@@ -0,0 +1,633 @@
...
+            {
+                return Comparer.Compare(x, y) <= 0;
+            }
+
+        }
+
+        /// <summary>
+        /// Comparator which sorts Vbox instances based on their volume, in descending order
+        /// </summary>
+        private sealed class VBOX_COMPARATOR_VOLUME : Comparer<Vbox>
```
Comment (JoogabYun) [inline src/Tizen.NUI/src/internal/Utility/ColorCutQuantizer.cs:378]: Please do not use "_" at class name.

"VBOX_COMPARATOR_VOLUME" -> "VboxComparatorVolume"

File: test/Tizen.NUI.Samples/Tizen.NUI.Samples/Samples/PaletteSample.cs [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+using System.Collections.Generic;
+using System.Diagnostics;
+using Tizen.NUI.BaseComponents;
+using Tizen.NUI.Components;
+
+namespace Tizen.NUI.Samples
+{
+    class PaletteSample : IExample
+    {
+        private static int BOTTOM_HEIGHT = 60;
```
Comment (JoogabYun) [inline test/Tizen.NUI.Samples/Tizen.NUI.Samples/Samples/PaletteSample.cs:30]: Please do not use "_" at naming.

'BOTTOM_HEIGHT' -> 'bottomHeight'

File: test/Tizen.NUI.Samples/Tizen.NUI.Samples/Samples/PaletteSample.cs [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+        private static int BUTTON_HEIGHT = 40;
+        private static int MAX_VIEW = 2;
+        private static string ResourcePath = Tizen.Applications.Application.Current.DirectoryInfo.Resource;
+        private static string[] IMAGE_PATH = {
+            ResourcePath + "/images/PaletteTest/rock.jpg",
+            ResourcePath + "/images/PaletteTest/red2.jpg",
+            ResourcePath + "/images/PaletteTest/10by10tree.png"
+        };
+
+        private Window CurrentWindow;
```
Comment (JoogabYun) [inline test/Tizen.NUI.Samples/Tizen.NUI.Samples/Samples/PaletteSample.cs:41]: 'CurrentWindow' -> 'currentWindow'

File: test/Tizen.NUI.Samples/Tizen.NUI.Samples/Samples/PaletteSample.cs [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+        private int ViewIndex = 0;
+        private int WindowWidth, WindowHeight;
+        private View View;
+        private View BottomView;
+        private ImageView ImageView;
+        private Palette P;
```
Comment (JoogabYun) [inline test/Tizen.NUI.Samples/Tizen.NUI.Samples/Samples/PaletteSample.cs:47]: please use a meaningful name.

'P' -> 'palette' etc..

**Comment by lwc0917** [comment]
@JoogabYun Sorry for the coding rule miss.
This time, I modified it without using the automatic tool. :)

**Review by JoogabYun — APPROVED** [review]

**Pull request merged**